### PR TITLE
Honor id/UUID when posting permission set + refactor MODPERMS-84

### PR DESCRIPTION
**Pull request by adamdickmeiss**
Fixes the issue of MODPERMS-84. It also returns 400 error where it previously returned 500 in a number of cases where user input . eg tenant is bad and fails due to bad user input. Not an internal error.. The 500 error handling has been re-done.. and so the coverage is low for this PR. Still overall coverage is up from 63.7 to 71.1.

### CODE REVIEW

File: src/main/java/org/folio/rest/impl/PermsAPI.java [hunk 1]
```diff
@@ -140,51 +139,24 @@ private static void report(String noise, Logger logger) {
 
   @Override
   public void getPermsUsers(int length, int start, String sortBy, String query,
```
Comment (julianladisch) [inline src/main/java/org/folio/rest/impl/PermsAPI.java:141]: Can you add `@Validate` here and for all other methods?

File: src/main/java/org/folio/rest/impl/PermsAPI.java [hunk 2]
```diff
@@ -613,20 +534,18 @@ public void postPermsUsersPermissionsById(String id, String indexField,
...
       PostgresClient.getInstance(vertxContext.owner(), tenantId).get(
           TABLE_NAME_PERMSUSERS,
           PermissionUser.class, new Criterion(useridCrit), true, false,
           getReply -> {
             try {
               if (getReply.failed()) {
-                logger.error("Error checking for user: " + getReply.cause()
-                    .getMessage(), getReply.cause());
+                String errStr = "Error checking for user: " + getReply.cause().getMessage();
+                logger.error(errStr);
```
Comment (julianladisch) [inline src/main/java/org/folio/rest/impl/PermsAPI.java:546]:
```suggestion
                logger.error(errStr, getReply.cause());
```

File: src/main/java/org/folio/rest/impl/PermsAPI.java [hunk 3]
```diff
@@ -763,14 +677,15 @@ public void deletePermsUsersPermissionsByIdAndPermissionname(
...
     try {
-      String tenantId = TenantTool.calculateTenantId(okapiHeaders.get(OKAPI_TENANT_HEADER));
+      String tenantId = TenantTool.tenantId(okapiHeaders);
       Criteria idCrit = getIdCriteria(indexField, "=", id);
       PostgresClient.getInstance(vertxContext.owner(), tenantId).get(TABLE_NAME_PERMSUSERS,
           PermissionUser.class, new Criterion(idCrit), true, false, getReply -> {
             if (getReply.failed()) {
-              logger.error("Error checking for user: " + getReply.cause().getMessage());
+              String errStr = getReply.cause().getMessage();
+              logger.error(errStr);
```
Comment (julianladisch) [inline src/main/java/org/folio/rest/impl/PermsAPI.java:686]:
```suggestion
              logger.error(errStr, getReply.cause());
```

File: src/main/java/org/folio/rest/impl/PermsAPI.java [hunk 4]
```diff
@@ -859,8 +778,9 @@ public void postPermsPermissions(PermissionUpload entity, Map<String, String> ok
       PostgresClient.getInstance(vertxContext.owner(), tenantId).get(
           TABLE_NAME_PERMS, Permission.class, new Criterion(nameCrit), true, false, getReply -> {
             if (getReply.failed()) {
-              logger.error("Error getting existing permissions: " + getReply.cause().getMessage());
-              asyncResultHandler.handle(Future.succeededFuture(PostPermsPermissionsResponse.respond500WithTextPlain("Internal server error")));
+              String errStr = getReply.cause().getMessage();
+              logger.error(errStr);
```
Comment (julianladisch) [inline src/main/java/org/folio/rest/impl/PermsAPI.java:782]:
```suggestion
              logger.error(errStr, getReply.cause());
```

File: src/main/java/org/folio/rest/impl/PermsAPI.java [hunk 5]
```diff
@@ -892,7 +813,9 @@ public void postPermsPermissions(PermissionUpload entity, Map<String, String> ok
                   if (postReply.failed()) {
                     postgresClient.rollbackTx(connection, done -> {
                       logger.error("Unable to save new permission: " + postReply.cause().getMessage());
```
Comment (julianladisch) [inline src/main/java/org/folio/rest/impl/PermsAPI.java:815]:
```suggestion
                      logger.error("Unable to save new permission: " + postReply.cause().getMessage(), postReply.cause());
```

File: src/main/java/org/folio/rest/impl/PermsAPI.java [hunk 6]
```diff
@@ -938,33 +859,33 @@ public void getPermsPermissionsById(String id, Map<String, String> okapiHeaders,
...
       idCrit.addField(ID_FIELD);
       idCrit.setOperation("=");
       idCrit.setVal(id);
       PostgresClient.getInstance(vertxContext.owner(), tenantId)
           .get(TABLE_NAME_PERMS, Permission.class, new Criterion(idCrit), true, false, getReply -> {
             if (getReply.failed()) {
-              logger.error("Error in getReply: " + getReply.cause().getMessage());
-              asyncResultHandler.handle(Future.succeededFuture(GetPermsPermissionsByIdResponse.respond500WithTextPlain("Internal server error")));
+              String errStr = getReply.cause().getMessage();
+              logger.error(errStr);
```
Comment (julianladisch) [inline src/main/java/org/folio/rest/impl/PermsAPI.java:871]:
```suggestion
              logger.error(errStr, getReply.cause());
```

File: src/main/java/org/folio/rest/impl/PermsAPI.java [hunk 7]
```diff
@@ -223,151 +195,120 @@ public void postPermsUsers(PermissionUser entity, Map<String, String> okapiHeade
...
-      String tenantId = TenantTool.calculateTenantId(okapiHeaders.get(OKAPI_TENANT_HEADER));
-      Criteria idCrit = getIdCriteria(indexField, "=", decodedId);
+      String tenantId = TenantTool.tenantId(okapiHeaders);
+      Criteria idCrit = getIdCriteria(indexField, "=", id);
       PostgresClient.getInstance(vertxContext.owner(), tenantId).get(TABLE_NAME_PERMSUSERS, PermissionUser.class,
           new Criterion(idCrit), true, false, queryReply -> {
             if (queryReply.failed()) {
-              String errStr = "queryReply failed: " + queryReply.cause().getMessage();
+              String errStr = queryReply.cause().getMessage();
               logger.error(errStr);
```
Comment (julianladisch) [inline src/main/java/org/folio/rest/impl/PermsAPI.java:257]:
```suggestion
              logger.error(errStr, queryReply.cause());
```

File: src/test/java/org/folio/permstest/RestVerticleTest.java [hunk 8]
```diff
@@ -696,34 +980,21 @@ public void testGroup(TestContext context) throws InterruptedException, Executio
 
   private void send(String url, TestContext context, HttpMethod method, String content,
                     String contentType, Handler<HttpClientResponse> handler) {
```
Comment (julianladisch) [inline src/test/java/org/folio/permstest/RestVerticleTest.java:982]: Duplicated code should be extracted into methods for get, post, put, delete. A few examples:
```
private Response send(String tenant, HttpMethod method, String path, String content, TestContext context) {
  try {
    CompletableFuture<Response> futureResponse = new CompletableFuture();
    send(tenant, "http://localhost:" + port + path, context, method, content,
          SUPPORTED_CONTENT_TYPE_JSON_DEF, new HTTPResponseHandler(futureResponse));
    return futureResponse.get(5, TimeUnit.SECONDS);
  } catch (Exception e) {
    context.fail(e);
    return null;
  }
}

private Response get(String path, TestContext context) {
  return send("diku", HttpMethod.GET, path, null, context);
}

private Response get(String tenant, String path, TestContext context) {
  return send(tenant, HttpMethod.GET, path, null, context);
}

private Response put(String path, JsonObject content, TestContext context) {
  return send("diku", HttpMethod.PUT, path, content.encode(), context);
}

private Response put(String path, String content, TestContext context) {
  return send("diku", HttpMethod.PUT, path, content, context);
}

private Response put(String tenant, String path, JsonObject content, TestContext context) {
  return send(tenant, HttpMethod.PUT, path, content.encode(), context);
}

```

Comment (adamdickmeiss) [reply src/test/java/org/folio/permstest/RestVerticleTest.java:982]: Yes. It's quite bad.. Something like restassured would be better. Can we postpone this to MODINVSTOR-83?

Comment (julianladisch) [reply src/test/java/org/folio/permstest/RestVerticleTest.java:982]: Yes, postpone to MODPERMS-83. I don't think that restassured would be better.

**Pull request merged**